crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+            Ok(MultiEd25519PrivateKey(private_keys, threshold))
+        } else {
+            Err(CryptoMaterialError::WrongLengthError)
+        }
+    }
+
+    /// Serialize a MultiEd25519PrivateKey.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        to_bytes(&self.0, self.1)
+    }
+}
+
+impl MultiEd25519PublicKey {
+    /// Construct a new MultiEd25519PublicKey.
+    /// --- Rules ---
+    /// a) threshold cannot be zero.
+    /// b) public_keys.len() should be equal to or larger than threshold.
+    /// c) support up to MAX_NUM_OF_KEYS public keys.
+    pub fn new(
+        public_keys: Vec<Ed25519PublicKey>,
+        threshold: u8,
+    ) -> std::result::Result<Self, CryptoMaterialError> {
+        let num_of_keys = public_keys.len();
+        if threshold != 0 && num_of_keys >= threshold as usize && num_of_keys <= MAX_NUM_OF_KEYS {
+            Ok(MultiEd25519PublicKey(public_keys, threshold))
+        } else {
+            Err(CryptoMaterialError::WrongLengthError)
+        }
+    }
+
+    /// Serialize a MultiEd25519PublicKey.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        to_bytes(&self.0, self.1)
+    }
+}
+
+///////////////////////
+// PrivateKey Traits //
+///////////////////////
+
+impl PrivateKey for MultiEd25519PrivateKey {
+    type PublicKeyMaterial = MultiEd25519PublicKey;
+}
+
+impl SigningKey for MultiEd25519PrivateKey {
+    type VerifyingKeyMaterial = MultiEd25519PublicKey;
+    type SignatureMaterial = MultiEd25519Signature;
+
+    // Sign a message with the minimum amount of keys to meet threshold (starting from left-most keys).
+    fn sign_message(&self, message: &HashValue) -> MultiEd25519Signature {
+        let mut ed25519_signatures: Vec<Ed25519Signature> = Vec::with_capacity(self.1 as usize);
+        let mut bitvec = BitVec::from_elem(MAX_NUM_OF_KEYS, false);
+
+        for (i, item) in self.0.iter().enumerate() {
+            if i < self.1 as usize {
+                ed25519_signatures.push(item.sign_message(message));
+                bitvec.set(i, true);
+            } else {
+                break;
+            }
+        }
+        MultiEd25519Signature(ed25519_signatures, bitvec)
+    }
+}
+
+// Generating a random 2 out-of 3 key for testing.
+impl Uniform for MultiEd25519PrivateKey {
+    fn generate_for_testing<R>(rng: &mut R) -> Self
+    where
+        R: ::rand::SeedableRng + ::rand::RngCore + ::rand::CryptoRng,
+    {
+        let mut private_keys: Vec<Ed25519PrivateKey> = Vec::with_capacity(3);
+        for _ in 0..3 {
+            private_keys.push(
+                Ed25519PrivateKey::try_from(
+                    &ed25519_dalek::SecretKey::generate(rng).to_bytes()[..],
+                )
+                .unwrap(),
+            );
+        }
+        MultiEd25519PrivateKey(private_keys, 2)
+    }
+}
+
+impl PartialEq<Self> for MultiEd25519PrivateKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.to_bytes() == other.to_bytes()
+    }
```

Comment:
```suggestion
    fn eq(&self, other: &Self) -> bool {
        self.1 == other.1
            && self.0.len() == other.0.len()
            && self
                .0
                .iter()
                .zip(&other.0)
                .all(|(left, right)| left == right)
    }
```
(this short-cuts, see https://github.com/libra/libra/pull/2723/files#r385427170)